pkg/registry/apps/statefulset/strategy_test.go

```diff
@@ -271,6 +272,180 @@ func TestStatefulSetStrategy(t *testing.T) {
+		}
+
+		// update validation should allow maxUnavailable changes
+		oldValidMU := validPercentMU.DeepCopy()
+		oldValidMU.ObjectMeta = metav1.ObjectMeta{ResourceVersion: "1", Generation: 1}
+
+		newValidMU := oldValidMU.DeepCopy()
+		newValidMU.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable = &intstr.IntOrString{Type: intstr.String, StrVal: "50%"}
+
+		Strategy.PrepareForUpdate(ctx, newValidMU, oldValidMU)
+		errs = Strategy.ValidateUpdate(ctx, newValidMU, oldValidMU)
+		if len(errs) != 0 {
+			t.Errorf("updating MaxUnavailable should be allowed on a statefulset: %v", errs)
+		}
+
+		// maxUnavailable with OnDelete strategy should not be allowed
+		onDeleteMU := &apps.StatefulSet{
+			ObjectMeta: metav1.ObjectMeta{Name: "test-ondelete", Namespace: metav1.NamespaceDefault},
+			Spec: apps.StatefulSetSpec{
+				PodManagementPolicy: apps.OrderedReadyPodManagement,
+				Selector:            &metav1.LabelSelector{MatchLabels: validSelector},
+				Template:            validPodTemplate.Template,
+				UpdateStrategy: apps.StatefulSetUpdateStrategy{
+					Type: apps.OnDeleteStatefulSetStrategyType,
+					RollingUpdate: &apps.RollingUpdateStatefulSetStrategy{
+						MaxUnavailable: &intstr.IntOrString{Type: intstr.Int, IntVal: 1},
+					},
+				},
+			},
+		}
+		Strategy.PrepareForCreate(ctx, onDeleteMU)
+		errs = Strategy.Validate(ctx, onDeleteMU)
+		if len(errs) == 0 {
+			t.Errorf("expected failure when MaxUnavailable is used with OnDelete strategy but got no error")
+		}
+	})
+
+	t.Run("StatefulSet maxUnavailable feature gate disabled behavior", func(t *testing.T) {
+		featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.MaxUnavailableStatefulSet, false)
+
+		// creation with maxUnavailable when feature gate is disabled
+		maxUnavailableMU := &apps.StatefulSet{
+			ObjectMeta: metav1.ObjectMeta{Name: "test-feature-disabled", Namespace: metav1.NamespaceDefault},
+			Spec: apps.StatefulSetSpec{
+				PodManagementPolicy: apps.ParallelPodManagement,
+				Selector:            &metav1.LabelSelector{MatchLabels: validSelector},
+				Template:            validPodTemplate.Template,
+				UpdateStrategy: apps.StatefulSetUpdateStrategy{
+					Type: apps.RollingUpdateStatefulSetStrategyType,
+					RollingUpdate: &apps.RollingUpdateStatefulSetStrategy{
+						MaxUnavailable: &intstr.IntOrString{Type: intstr.Int, IntVal: 2},
+					},
+				},
+			},
+		}
+		Strategy.PrepareForCreate(ctx, maxUnavailableMU)
+
+		// The field should be dropped during prepare
+		if maxUnavailableMU.Spec.UpdateStrategy.RollingUpdate != nil && maxUnavailableMU.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable != nil {
+			t.Errorf("expected MaxUnavailable to be dropped when feature gate is disabled")
+		}
+
+		errs := Strategy.Validate(ctx, maxUnavailableMU)
+		if len(errs) != 0 {
+			t.Errorf("unexpected error validating StatefulSet when MaxUnavailable is dropped: %v", errs)
+		}
+	})
```

Comment:
All of the above is already covered in [validation tests](https://github.com/kubernetes/kubernetes/blob/091f87c10bc3532041b77a783a5f832de5506dc8/pkg/apis/apps/validation/validation_test.go#L702-L730) so I'm not sure we need this here. If you want to improve, I'd accept a PR for that validation tests covering cases that are not covered, such as negative maxUnavailable. Although I'm inclined to say you can just add it to [your beta PR](https://github.com/kubernetes/kubernetes/pull/133153).